feat: Add converting prps to prpd

hjkim-onepredict opened the pull request:
## Summary
_Write the summary about this PR._
First commit for this issue.
Make function and test script.

This function convert PRPS data, which represents the amplitude of partial discharge in time and phase dimension to PRPD data by projecting in time dimension.
- 

## Changes
_Write changes in detail._

- 

## How to Test?
_Write a way to test the changes._

- 

## Checklist

- [x] Do you pass all the tests?
- [x] Do you write the changes on document such as README?

Commit:
feat: 작업물을 다른 환경으로 옮기기 위함 (#53)

[-- body --]

Commit:
feature: Initial commit(#53)

[-- body --]

kyunghwan-onepredict commented:
해당 PR에 대한 간단한 설명 작성 부탁드립니다.

kyunghwan-onepredict commented on onebone/preprocessing/pd.py:
```diff
@@ -0,0 +1,41 @@
+
+def ps2pd(ps, range_amp: Tuple[int, int] = (0, 256), resol_amp: int = 128) -> np.ndarray:
+    """
+    Transform prps(phase resolved pulse sequance) to a prpd(phaes resolved partial discharge) by marginalizing time dimension.
```
오타 수정
```suggestion
    Transform prps(phase resolved pulse sequence) to a prpd(phaes resolved partial discharge) by marginalizing time dimension.
```

isingmodel replied:
phaes도 오타네요 ㅋㅋ

kyunghwan-onepredict commented on onebone/preprocessing/pd.py:
```diff
@@ -0,0 +1,41 @@
+        The data. Ex: kepco standard=(128, 3600)
+    range_amp : tuple (min, max), default=(0, 256)
+        Measurement range of PD DAQ. Refers to DAQ manufacture.
+    resol_amp : int, default 128
```
오타 수정
```suggestion
    resol_amp : int, default=128
```

hjkim-onepredict replied:
감사합니다.

kyunghwan-onepredict commented on tests/preprocessing/test_pd.py:
```diff
@@ -0,0 +1,34 @@
+"""Test code for pd.py
```
오타 수정
```suggestion
"""Test code for pd.py.
```

kyunghwan-onepredict commented:
로컬에서 `make test` 실행하시고 에러 해결 필요합니다.

isingmodel commented on onebone/preprocessing/pd.py:
```diff
@@ -0,0 +1,41 @@
+"""Transform PRPS format pd data to PRPD format.
+
+- Author: Hyunjae Kim
+- Contact: hyunjae.kim@onepredict.com
```
preprocessing/pd.py가 적절한 카테고리명인지 잘 모르겠습니다. 좀 더 일반적인 이름 없을까요?

hjkim-onepredict replied:
PRPS를 전처리하여 PRPD로 바꾸는 것이라 일단 preprocessing module에 넣었고, pd 관련된 전처리를 앞으로 더 추가해볼 예정입니다.

kyunghwan-onepredict replied:
@hjkim-onepredict `gis_pd` 정도로 이름 짓는건 어떨까요?

isingmodel commented on tests/preprocessing/test_pd.py:
```diff
@@ -0,0 +1,34 @@
+    prps, prpd = _generate_pspd(coord, cum_value)
+    prpd_transformed = pd.ps2pd(prps)
+    # Check the output
+    assert (prpd_transformed == prpd).all()
```
np.zeros를 사용하면 알고리즘의 동작 여부와 상관없이 차원만 맞으면 test 통과가 되는 것이니 test 기준이 약한 것 아닐까요?

hjkim-onepredict replied:
네 맞습니다. 엄밀한 테스트코드를 업데이트해서 다시 커밋하도록 하겠습니다.

Kangwhi-Kim commented on onebone/preprocessing/pd.py:
```diff
@@ -0,0 +1,41 @@
+    bins = np.linspace(range_amp[0], range_amp[1], resol_amp)
+    p_digit = np.digitize(ps, bins) - 1
+    p_digit_logit = imat[p_digit]
+    pd = np.sum(p_digit_logit, axis=0)
```
`ps`가 1-d array이면 `np.sum(..., axis=0)`을 할 필요가 없을 것 같습니다.
```suggestion
    if len(p_digit_logit.shape) == 1:
        pd = p_digit_logit
    else:
        pd = np.sum(p_digit_logit, axis=0)
```
`ps`를 2차원으로 의도하신 것 같은데 이렇게도 쓸 수 있을것 같습니다.
```suggestion
    if len(p_digit_logit.shape) == 1:
        raise ValueError('`ps` has to be 2-dimensions.')
```

hjkim-onepredict replied:
PRPS의 데이터 자체가 첫번째 차원이 phase, 두번째 차원이 time 이기 때문에 무조건 2차원입니다. 컬러 이미지가 3차원으로 정의되는것처럼 해당 도메인에서의 상식으로 생각해주시면 됩니다.

kyunghwan-onepredict replied:
강휘님이 아래에 제안주신대로 입력이 2차원이 아닐 경우 error를 내는 제약조건이 있으면 더 좋을것 같습니다.

Kangwhi-Kim replied:
> PRPS의 데이터 자체가 첫번째 차원이 phase, 두번째 차원이 time 이기 때문에 무조건 2차원입니다. 컬러 이미지가 3차원으로 정의되는것처럼 해당 도메인에서의 상식으로 생각해주시면 됩니다.

그렇군요, 그러면 `ps`에 대해 2차원 array만 입력될 수 있도록 제약조건을 두는 것이 좋을 것 같습니다.

hjkim-onepredict replied:
두분 감사합니다. 의견 반영하여 수정하였습니다.

Kangwhi-Kim commented on tests/preprocessing/test_pd.py:
```diff
@@ -0,0 +1,34 @@
+
+
+def _generate_pspd(coord_point: Tuple[int, int], cum_value: int) -> Tuple[np.ndarray, np.ndarray]:
+    prps = np.zeros([3600, 128])
+    prpd = np.zeros([128, 128])
+    return prps, prpd
```
`coord_point`와 `cum_value` 인자가 사용되지 않았습니다. `ps2pd` 테스트에 사용하시려고 선언하신 것 같은데 업데이트 요청드립니다.

hjkim-onepredict replied:
테스트용 샘플 PRPS, PRPD를 generate하기 위한 인자이며, generate 및 비교코드를 업데이트하도록 하겠습니다.

hjkim-onepredict replied:
테스트 코드 작성하여 커밋하였습니다.

Kangwhi-Kim commented on onebone/preprocessing/pd.py:
```diff
@@ -0,0 +1,41 @@
+    p_digit = np.digitize(ps, bins) - 1
+    p_digit_logit = imat[p_digit]
+    pd = np.sum(p_digit_logit, axis=0)
+    pd = np.fliplr(pd).transpose()
```
단순 transpose만 하는게 아니라 이에 더해 `np.fliplr`을 사용한 목적이 있으신가요?

hjkim-onepredict replied:
마지막에 좌우를 뒤집어주지 않으면 PRPD를 이미지로 표현할 때 phase가 역순으로 보이기 때문입니다.

Kangwhi-Kim replied:
그렇군요, 감사합니다.

hjkim-onepredict committed:
Update tests/preprocessing/test_pd.py

맞습니다. 감사합니다.

Co-authored-by: Kyunghwan Kim <86699249+kyunghwan-onepredict@users.noreply.github.com>

hjkim-onepredict committed:
Update onebone/preprocessing/pd.py

Co-authored-by: Kyunghwan Kim <86699249+kyunghwan-onepredict@users.noreply.github.com>

hjkim-onepredict committed:
Update onebone/preprocessing/pd.py

Co-authored-by: Kyunghwan Kim <86699249+kyunghwan-onepredict@users.noreply.github.com>

kyunghwan-onepredict commented on onebone/preprocessing/pd.py:
```diff
@@ -0,0 +1,41 @@
+"""Transform PRPS format pd data to PRPD format.
```
Full name을 언급해주는게 이해에 도움이 될것 같네요
- PRPD(Phase Resolved Partial Discharge)
- PRPS(Phase Resolved Pulse Sequence)

hjkim-onepredict replied:
수정하도록 하겠습니다.

hjkim-onepredict replied:
수정하였습니다!

Commit:
[feature]: convert prps to prpd(#83)

Add error raise statement for unappropreate dimension of input.

Commit:
Merge branch 'feature/convert-prps-to-prpd' of https://github.com/Onepredict/onebone into feature/convert-prps-to-prpd

Commit:
feature: convert prps to prpd #83

add code use example.

Commit:
feature: convert prps to prpd(#53)

[-- body --]

Commit:
feature: convert prps to prpd(#53)

add test code. this code generate arbitrary prps and prpd that
represents partial discharge signal arising only from a ramdom pair of phase
and amplitude. And compare the transformed prpd using generated prps to
original prpd. This test iterates 10 times.

hjkim-onepredict commented on onebone/preprocessing/pd.py:
```diff
@@ -0,0 +1,41 @@
+
+    Examples
+    --------
+    To be filled.
```
추가하였습니다. 입출력이 너무 단순해서 그림 같은걸 삽입해서 표현하고 싶은데 나중에 버전업 때 반영하여도 괜찮을까요?

Commit:
feature: convert prps to prpd(#53)

edit discription of the script pd.py

hjkim-onepredict commented:
> 로컬에서 `make test` 실행하시고 에러 해결 필요합니다.

Commit:
feature: convert prps to prpd(#53)

Edit test code. Previous commit was done before make test by mistake.
Sorry.

Commit:
feature: convert prps to prpd(#53)

Add random condition generation part for the iterative test.

kyunghwan-onepredict commented on onebone/preprocessing/pd.py:
```diff
@@ -0,0 +1,50 @@
+        ...,
+        [0., 0., 0., ..., 0., 0., 0.]])
+    """
+    if len(ps.shape) == 1:
```
```suggestion
    if len(ps.shape) != 2:
```
이게 더 명확할 것 같습니다.

kyunghwan-onepredict commented on tests/preprocessing/test_pd.py:
```diff
@@ -0,0 +1,47 @@
+
+
+def test_ps2pd():
+    for _ in range(10):
```
seed를 걸어놓고 한번만 테스트해도 괜찮을것 같습니다.

hjkim-onepredict replied:
해당 부분 수정완료하였씁니다.

kyunghwan-onepredict commented:
@hjkim-onepredict CI test 처리 부탁드려요.

hjkim-onepredict committed:
Update onebone/preprocessing/pd.py

Co-authored-by: Kyunghwan Kim <86699249+kyunghwan-onepredict@users.noreply.github.com>

kyunghwan-onepredict commented on onebone/preprocessing/pd.py:
```diff
@@ -0,0 +1,50 @@
+
+- Author: Hyunjae Kim
+- Contact: hyunjae.kim@onepredict.com
+"""
```
@hjkim-onepredict gis_pd 정도로 이름 짓는건 어떨까요?

Kangwhi-Kim replied:
gis가 무슨 약자인가요? ㅋㅋ

kyunghwan-onepredict replied:
설비 이름 같습니다. [참고](https://m.blog.naver.com/PostView.naver?isHttpsRedirect=true&blogId=k8010314&logNo=221034364787)

hjkim-onepredict replied:
pd 알고리즘을 모터 등 다른 고전압 설비에도 활용할 계획이 있다고 들어서, 일단 pd로 가보는게 어떨까 합니다.

kyunghwan-onepredict commented on onebone/preprocessing/pd.py:
```diff
@@ -0,0 +1,50 @@
+
+    Examples
+    --------
+    >>> ps = np.random.random([3600,128])
```
- 의미있는 값으로 example을 보여주기 어려운 상황인것 같은데 맞나요?
- 함수를 통과하면 shape가 변하게 되니 그걸 example로 하는게 어떨까요?

Commit:
feature: Convert prps to prpd(#53)

Fix length of comments

Commit:
Merge branch 'feature/convert-prps-to-prpd' of https://github.com/Onepredict/onebone into feature/convert-prps-to-prpd

Commit:
feature: convert prps to prpd (#53)

Edit range of test condition and fix seed.

Commit:
feature: convert prps to prps(#53)

Fix random seed.

Commit:
feature: convert prps to prpd(#53)

Remove for loop foro test iteration

Kangwhi-Kim reviewed (approved):
고생많으셨습니다 현재님!
기여해주셔서 감사합니다. ㅎㅎ

kyunghwan-onepredict reviewed (approved):
Excellent!!

kyunghwan-onepredict commented:
@all-contributors please add @hjkim-onepredict for code

kyunghwan-onepredict committed:
Merge branch 'main' into feature/convert-prps-to-prpd

kyunghwan-onepredict commented:
@all-contributors please add @hjkim-onepredict for code

kyunghwan-onepredict commented:
@all-contributors please remove @hjkim-onepredict for code

allcontributors[bot] committed:
docs: add hjkim-onepredict as a contributor for code (#94)

* docs: update README.md [skip ci]

* docs: update .all-contributorsrc [skip ci]

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>
Co-authored-by: Kyunghwan Kim <kyunghwan.kim@onepredict.com>
